src/imp_pl.rs

```diff
@@ -56,13 +57,30 @@ impl<T> OnceCell<T> {
             // - finally, if it returns Ok, we store the value and store the flag with
             //   `Release`, which synchronizes with `Acquire`s.
             let value = f()?;
-            let slot: &mut Option<T> = unsafe { &mut *self.value.get() };
-            debug_assert!(slot.is_none());
-            *slot = Some(value);
+            unsafe { self.value.get().write(MaybeUninit::new(value)) };
             self.is_initialized.store(true, Ordering::Release);
         }
         Ok(())
     }
+
+    /// Marks this cell as uninitialized.
+    ///
+    /// # Safety
+    ///
+    /// It's not intended for concurrent use, that's why using
+    /// `Ordering::Relaxed` and accepting `&self` made for performance here.
+    ///
+    /// It's up to the caller to guarantee that:
+    /// - this function is not called concurrently (should be used where
+    ///   `&mut self` or `self` is accepted);
+    /// - the value contained in this cell is properly moved out before calling
+    ///   this function.
+    ///
+    /// Violating the rules above will cause UB (undefined behavior).
```

Comment:
@WaffleLapkin

> so what's the point of accepting `&self` (instead `&mut self`)

I've thought about possible better compiler optimizations.

> And the second case doesn't seem to be UB at all. If you think that you have uninitialized value, but actually you have an initialized value, then it could lead to memory leak, but not UB.

But what about overwritten value not being properly dropped?

Reply:
> I've thought about possible better compiler optimizations.

But with `&mut` compiler knows more and so could optimize better, doesn't it?

> But what about overwritten value not being properly dropped?

Forgetting to drop is not **UB**, it's only (a possible) memory leak. See [mem::forget](https://doc.rust-lang.org/std/mem/fn.forget.html)

Reply:
@WaffleLapkin makes sense. Thanks for clarifying this!